File: sdks/go/registry.go
package mdp

import (
	"errors"
	"fmt"
	"strings"
	"sync"
)

type PathHandler func(PathRequest, PathInvocationContext) (any, error)

type procedureEntry struct {
	descriptor PathDescriptor
	handler    PathHandler
}

type resolvedProcedure struct {
	descriptor  PathDescriptor
	handler     PathHandler
	params      map[string]any
	specificity []int
}

type ProcedureRegistry struct {
	mu      sync.RWMutex
	entries []procedureEntry
}

func NewProcedureRegistry() *ProcedureRegistry {
	return &ProcedureRegistry{}
}

func (registry *ProcedureRegistry) ExposeEndpoint(
	path string,
	method HttpMethod,
	handler PathHandler,
	options EndpointOptions,
) error {
	return registry.register(
		EndpointPathDescriptor{
			PathValue:    path,
			Method:       method,
			Description:  options.Description,
			InputSchema:  options.InputSchema,
			OutputSchema: options.OutputSchema,
			ContentType:  options.ContentType,
		},
		handler,
	)
}

func (registry *ProcedureRegistry) ExposeSkillMarkdown(
	path string,
	content string,
	options SkillOptions,
) error {
	if options.ContentType == "" {
		options.ContentType = "text/markdown"
	}
	description := options.Description
	if description == "" {
		description = deriveMarkdownDescription(content)
	}
	return registry.register(
		SkillPathDescriptor{
			PathValue:   path,
			Description: description,
			ContentType: options.ContentType,
		},
		func(_ PathRequest, _ PathInvocationContext) (any, error) {
			return content, nil
		},
	)
}

func (registry *ProcedureRegistry) ExposePromptMarkdown(
	path string,
	content string,
	options PromptOptions,
) error {
	description := options.Description
	if description == "" {
		description = deriveMarkdownDescription(content)
	}
	return registry.register(
		PromptPathDescriptor{
			PathValue:    path,
			Description:  description,
			InputSchema:  options.InputSchema,
			OutputSchema: options.OutputSchema,
		},
		func(_ PathRequest, _ PathInvocationContext) (any, error) {
			return map[string]any{
				"messages": []map[string]any{
					{
						"role":    "user",
						"content": content,
					},
				},
			}, nil
		},
	)
}

func (registry *ProcedureRegistry) Unexpose(path string, method *HttpMethod) (bool, error) {
	if !IsPathPattern(path) {
		return false, fmt.Errorf("invalid path %q", path)
	}

	registry.mu.Lock()
	defer registry.mu.Unlock()

	for index, entry := range registry.entries {
		if entry.descriptor.Path() != path {
			continue
		}
		if endpoint, ok := entry.descriptor.(EndpointPathDescriptor); ok {
			if method != nil && endpoint.Method != *method {
				continue
			}
		}
		registry.entries = append(registry.entries[:index], registry.entries[index+1:]...)
		return true, nil
	}

	return false, nil
}

func (registry *ProcedureRegistry) DescribePaths() []PathDescriptor {
	registry.mu.RLock()
	defer registry.mu.RUnlock()

	result := make([]PathDescriptor, 0, len(registry.entries))
	for _, entry := range registry.entries {
		result = append(result, entry.descriptor)
	}
	return result
}

func (registry *ProcedureRegistry) Describe(client ClientInfo) ClientDescriptor {
	return NewClientDescriptor(client, registry.DescribePaths())
}

func (registry *ProcedureRegistry) Invoke(message CallClientMessage) (any, error) {
	entry := registry.resolveEntry(message.Method, message.Path)
	if entry == nil {
		return nil, fmt.Errorf("unknown path %q for method %q", message.Path, message.Method)
	}

	return entry.handler(
		PathRequest{
			Params:  entry.params,
			Queries: cloneMap(message.Query),
			Body:    message.Body,
			Headers: cloneStringMap(message.Headers),
		},
		PathInvocationContext{
			RequestID: message.RequestID,
			ClientID:  message.ClientID,
			PathType:  entry.descriptor.DescriptorType(),
			Method:    message.Method,
			Path:      message.Path,
			Auth:      message.Auth,
		},
	)
}

func (registry *ProcedureRegistry) register(descriptor PathDescriptor, handler PathHandler) error {
	if handler == nil {
		return errors.New("path handler is required")
	}
	if !IsPathPattern(descriptor.Path()) {
		return fmt.Errorf("invalid path %q", descriptor.Path())
	}
	if err := assertDescriptorPathShape(descriptor); err != nil {
		return err
	}

	registry.mu.Lock()
	defer registry.mu.Unlock()

	key := registrationKey(descriptor)
	entry := procedureEntry{descriptor: descriptor, handler: handler}
	for index, current := range registry.entries {
		if registrationKey(current.descriptor) == key {
			registry.entries[index] = entry
			return nil
		}
	}
	registry.entries = append(registry.entries, entry)
	return nil
}

func (registry *ProcedureRegistry) resolveEntry(method HttpMethod, path string) *resolvedProcedure {
	registry.mu.RLock()
	defer registry.mu.RUnlock()

	var bestMatch *resolvedProcedure
	for _, entry := range registry.entries {
		if !matchesMethod(entry.descriptor, method) {
			continue
		}
		match := MatchPathPattern(entry.descriptor.Path(), path)
		if match == nil {
			continue
		}
		if bestMatch == nil || ComparePathSpecificity(match.Specificity, bestMatch.specificity) > 0 {
			bestMatch = &resolvedProcedure{
				descriptor:  entry.descriptor,
				handler:     entry.handler,
				params:      cloneMap(match.Params),
				specificity: append([]int(nil), match.Specificity...),
			}
		}
	}
	return bestMatch
}

func registrationKey(descriptor PathDescriptor) string {
	if endpoint, ok := descriptor.(EndpointPathDescriptor); ok {
		return string(endpoint.Method) + " " + endpoint.PathValue
	}
	return descriptor.Path()
}

func matchesMethod(descriptor PathDescriptor, method HttpMethod) bool {
	if endpoint, ok := descriptor.(EndpointPathDescriptor); ok {
		return endpoint.Method == method
	}
	return method == HttpMethodGet
}

func assertDescriptorPathShape(descriptor PathDescriptor) error {
	path := descriptor.Path()

	switch descriptor.(type) {
	case EndpointPathDescriptor:
		if IsSkillPath(path) || IsPromptPath(path) {
			return errors.New("endpoint path cannot target reserved skill or prompt leaves")
		}
	case SkillPathDescriptor:
		if !IsSkillPath(path) {
			return errors.New("skill paths must end with skill.md or SKILL.md")
		}
	case PromptPathDescriptor:
		if !IsPromptPath(path) {
			return errors.New("prompt paths must end with prompt.md or PROMPT.md")
		}
	}
	return nil
}

func deriveMarkdownDescription(content string) string {
	var paragraph []string
	for _, rawLine := range strings.Split(content, "\n") {
		line := strings.TrimSpace(rawLine)
		if line == "" {
			if len(paragraph) > 0 {
				break
			}
			continue
		}
		if strings.HasPrefix(line, "#") {
			continue
		}
		paragraph = append(paragraph, line)
	}
	return strings.Join(paragraph, " ")
}

func cloneMap(value map[string]any) map[string]any {
	if value == nil {
		return map[string]any{}
	}
	result := make(map[string]any, len(value))
	for key, entry := range value {
		result[key] = entry
	}
	return result
}

func cloneStringMap(value map[string]string) map[string]string {
	if value == nil {
		return map[string]string{}
	}
	result := make(map[string]string, len(value))
	for key, entry := range value {
		result[key] = entry
	}
	return result
}

File: sdks/rust/src/registry.rs
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

use async_trait::async_trait;
use serde_json::{json, Map, Value};

use crate::error::MdpClientError;
use crate::models::{
    ClientDescriptor, ClientInfo, EndpointOptions, HttpMethod, PathDescriptor, PathInvocationContext,
    PathRequest, PromptOptions, SkillOptions,
};
use crate::path_utils::{
    compare_path_specificity, is_path_pattern, is_prompt_path, is_skill_path, match_path_pattern,
};
use crate::protocol::CallClientRequest;

type HandlerFuture = Pin<Box<dyn Future<Output = Result<Value, MdpClientError>> + Send>>;

#[async_trait]
pub trait PathHandler: Send + Sync {
    async fn handle(
        &self,
        request: PathRequest,
        context: PathInvocationContext,
    ) -> Result<Value, MdpClientError>;
}

#[async_trait]
impl<F, Fut> PathHandler for F
where
    F: Send + Sync + 'static + Fn(PathRequest, PathInvocationContext) -> Fut,
    Fut: Future<Output = Result<Value, MdpClientError>> + Send + 'static,
{
    async fn handle(
        &self,
        request: PathRequest,
        context: PathInvocationContext,
    ) -> Result<Value, MdpClientError> {
        (self)(request, context).await
    }
}

#[derive(Clone)]
struct ProcedureEntry {
    descriptor: PathDescriptor,
    handler: Arc<dyn PathHandler>,
}

#[derive(Clone)]
struct ResolvedProcedure {
    descriptor: PathDescriptor,
    handler: Arc<dyn PathHandler>,
    params: Map<String, Value>,
    specificity: Vec<i32>,
}

#[derive(Default, Clone)]
pub struct ProcedureRegistry {
    entries: Vec<ProcedureEntry>,
}

impl ProcedureRegistry {
    pub fn expose_endpoint<H, Fut>(
        &mut self,
        path: impl Into<String>,
        method: HttpMethod,
        handler: H,
        options: EndpointOptions,
    ) -> Result<(), MdpClientError>
    where
        H: Send + Sync + 'static + Fn(PathRequest, PathInvocationContext) -> Fut,
        Fut: Future<Output = Result<Value, MdpClientError>> + Send + 'static,
    {
        let descriptor = PathDescriptor::Endpoint {
            path: path.into(),
            method,
            description: options.description,
            input_schema: options.input_schema,
            output_schema: options.output_schema,
            content_type: options.content_type,
        };
        self.register(descriptor, Arc::new(handler))
    }

    pub fn expose_skill_markdown(
        &mut self,
        path: impl Into<String>,
        content: impl Into<String>,
        options: SkillOptions,
    ) -> Result<(), MdpClientError> {
        let content = content.into();
        let description = options
            .description
            .or_else(|| derive_markdown_description(&content));
        let descriptor = PathDescriptor::Skill {
            path: path.into(),
            description,
            content_type: options.content_type,
        };
        self.register(
            descriptor,
            Arc::new(move |_request: PathRequest, _context: PathInvocationContext| {
                let content = content.clone();
                Box::pin(async move { Ok(Value::String(content)) }) as HandlerFuture
            }),
        )
    }

    pub fn expose_prompt_markdown(
        &mut self,
        path: impl Into<String>,
        content: impl Into<String>,
        options: PromptOptions,
    ) -> Result<(), MdpClientError> {
        let content = content.into();
        let description = options
            .description
            .or_else(|| derive_markdown_description(&content));
        let descriptor = PathDescriptor::Prompt {
            path: path.into(),
            description,
            input_schema: options.input_schema,
            output_schema: options.output_schema,
        };
        self.register(
            descriptor,
            Arc::new(move |_request: PathRequest, _context: PathInvocationContext| {
                let content = content.clone();
                Box::pin(async move { Ok(json!({"messages": [{"role": "user", "content": content}]})) })
                    as HandlerFuture
            }),
        )
    }

    pub fn describe_paths(&self) -> Vec<PathDescriptor> {
        self.entries.iter().map(|entry| entry.descriptor.clone()).collect()
    }

    pub fn describe(&self, client: &ClientInfo) -> ClientDescriptor {
        ClientDescriptor::from_info(client, self.describe_paths())
    }

    pub fn unexpose(&mut self, path: &str, method: Option<HttpMethod>) -> Result<bool, MdpClientError> {
        assert_path_pattern(path)?;
        let method = method.as_ref();
        if let Some(index) = self.entries.iter().position(|entry| match &entry.descriptor {
            PathDescriptor::Endpoint {
                path: descriptor_path,
                method: descriptor_method,
                ..
            } => descriptor_path == path && method.map(|value| value == descriptor_method).unwrap_or(false),
            PathDescriptor::Skill { path: descriptor_path, .. }
            | PathDescriptor::Prompt { path: descriptor_path, .. } => descriptor_path == path,
        }) {
            self.entries.remove(index);
            return Ok(true);
        }
        Ok(false)
    }

    pub async fn invoke(&self, message: &CallClientRequest) -> Result<Value, MdpClientError> {
        let entry = self
            .resolve_entry(&message.method, &message.path)
            .ok_or_else(|| MdpClientError::UnknownPath {
                path: message.path.clone(),
                method: message.method.as_str().to_string(),
            })?;

        let request = PathRequest {
            params: entry.params,
            queries: message.query.clone().unwrap_or_default(),
            body: message.body.clone(),
            headers: message.headers.clone().unwrap_or_default(),
        };
        let context = PathInvocationContext {
            request_id: message.request_id.clone(),
            client_id: message.client_id.clone(),
            path_type: entry.descriptor.descriptor_type().to_string(),
            method: message.method.clone(),
            path: message.path.clone(),
            auth: message.auth.clone(),
        };
        entry.handler.handle(request, context).await
    }

    fn register(&mut self, descriptor: PathDescriptor, handler: Arc<dyn PathHandler>) -> Result<(), MdpClientError> {
        assert_path_pattern(descriptor.path())?;
        assert_descriptor_path_shape(&descriptor)?;
        let key = registration_key(&descriptor);
        let entry = ProcedureEntry { descriptor, handler };
        if let Some(index) = self
            .entries
            .iter()
            .position(|current| registration_key(&current.descriptor) == key)
        {
            self.entries[index] = entry;
        } else {
            self.entries.push(entry);
        }
        Ok(())
    }

    fn resolve_entry(&self, method: &HttpMethod, path: &str) -> Option<ResolvedProcedure> {
        let mut best_match: Option<ResolvedProcedure> = None;
        for entry in &self.entries {
            if !matches_method(&entry.descriptor, method) {
                continue;
            }
            let Some(path_match) = match_path_pattern(entry.descriptor.path(), path) else {
                continue;
            };

            if best_match
                .as_ref()
                .map(|current| compare_path_specificity(&path_match.specificity, &current.specificity) > 0)
                .unwrap_or(true)
            {
                best_match = Some(ResolvedProcedure {
                    descriptor: entry.descriptor.clone(),
                    handler: entry.handler.clone(),
                    params: path_match.params,
                    specificity: path_match.specificity,
                });
            }
        }
        best_match
    }
}

fn registration_key(descriptor: &PathDescriptor) -> String {
    match descriptor {
        PathDescriptor::Endpoint { path, method, .. } => format!("{} {}", method.as_str(), path),
        PathDescriptor::Skill { path, .. } | PathDescriptor::Prompt { path, .. } => path.clone(),
    }
}

fn matches_method(descriptor: &PathDescriptor, method: &HttpMethod) -> bool {
    match descriptor {
        PathDescriptor::Endpoint {
            method: descriptor_method,
            ..
        } => descriptor_method == method,
        PathDescriptor::Skill { .. } | PathDescriptor::Prompt { .. } => *method == HttpMethod::Get,
    }
}

fn assert_path_pattern(path: &str) -> Result<(), MdpClientError> {
    if is_path_pattern(path) {
        Ok(())
    } else {
        Err(MdpClientError::InvalidPath(path.to_string()))
    }
}

fn assert_descriptor_path_shape(descriptor: &PathDescriptor) -> Result<(), MdpClientError> {
    match descriptor {
        PathDescriptor::Endpoint { path, .. } => {
            if is_skill_path(path) || is_prompt_path(path) {
                return Err(MdpClientError::InvalidPath(path.clone()));
            }
        }
        PathDescriptor::Skill { path, .. } => {
            if !is_skill_path(path) {
                return Err(MdpClientError::InvalidPath(path.clone()));
            }
        }
        PathDescriptor::Prompt { path, .. } => {
            if !is_prompt_path(path) {
                return Err(MdpClientError::InvalidPath(path.clone()));
            }
        }
    }
    Ok(())
}

fn derive_markdown_description(content: &str) -> Option<String> {
    let mut paragraph = Vec::new();
    for line in content.lines().map(str::trim) {
        if line.is_empty() {
            if !paragraph.is_empty() {
                break;
            }
            continue;
        }
        if line.starts_with('#') {
            continue;
        }
        paragraph.push(line);
    }
    if paragraph.is_empty() {
        None
    } else {
        Some(paragraph.join(" "))
    }
}
